src/Compilers/CSharp/Test/Semantic/Semantics/RefLocalsAndReturnsTests.cs

```diff
@@ -12,14 +12,703 @@ namespace Microsoft.CodeAnalysis.CSharp.UnitTests.Semantics
+        }
+
+        [Fact]
+        public void RefForeachErrorRecovery()
+        {
+            var comp = CreateCompilation(@"
+class C
+{
+    void M()
+    {
+        foreach (ref var x in )
+        {
+        }
+    }
+}");
+            comp.VerifyDiagnostics(
+                // (6,31): error CS1525: Invalid expression term ')'
+                //         foreach (ref var x in )
+                Diagnostic(ErrorCode.ERR_InvalidExprTerm, ")").WithArguments(")").WithLocation(6, 31));
+        }
+
+        [Fact]
+        public void RefForToReadonly()
+        {
+            var comp = CreateCompilation(@"
+class C
+{
+    void M(in int x)
+    {
+        for (ref int i = ref x; i < 0; i++) {}
+    }
+}");
+            comp.VerifyDiagnostics(
+                // (6,30): error CS8329: Cannot use variable 'in int' as a ref or out value because it is a readonly variable
+                //         for (ref int i = ref x; i < 0; i++) {}
+                Diagnostic(ErrorCode.ERR_RefReadonlyNotField, "x").WithArguments("variable", "in int").WithLocation(6, 30));
+        }
+
+        [Fact]
+        public void RefForOutOfScope()
+        {
+            var comp = CreateCompilation(@"
+class C
+{
+    ref int M()
+    {
+        int x = 0;
+        for (ref int i = ref x; i < 0; i++)
+        {
+            if (i == 0)
+            {
+                return ref i;
+            }
+        }
+        return ref (new int[1])[0];
+    }
+}");
+            comp.VerifyDiagnostics(
+                // (11,28): error CS8157: Cannot return 'i' by reference because it was initialized to a value that cannot be returned by reference
+                //                 return ref i;
+                Diagnostic(ErrorCode.ERR_RefReturnNonreturnableLocal, "i").WithArguments("i").WithLocation(11, 28));
+        }
+
+        [Fact]
+        public void RefForeachReadonly()
+        {
+            var comp = CreateCompilation(@"
+using System;
+class C
+{
+    void M()
+    {
+        foreach (ref var v in new int[0])
+        {
+        }
+        foreach (ref readonly var v in new int[0])
+        {
+        }
+        foreach (ref var v in new RefEnumerable())
+        {
+            Console.WriteLine(v);
+        }
+    }
+}
+
+class RefEnumerable
+{
+    private readonly int[] _arr = new int[5];
+    public StructEnum GetEnumerator() => new StructEnum(_arr);
+
+    public struct StructEnum
+    {
+        private readonly int[] _arr;
+        private int _current;
+        public StructEnum(int[] arr)
+        {
+            _arr = arr;
+            _current = -1;
+        }
+        public ref readonly int Current => ref _arr[_current];
+        public bool MoveNext() => ++_current != _arr.Length;
+    }
+}");
+            comp.VerifyDiagnostics(
+                // (7,31): error CS1510: A ref or out value must be an assignable variable
+                //         foreach (ref var v in new int[0])
```

Comment:
Why? it doesn't work for `var x = new int[0]; foreach (ref var v in x)` either.

Reply:
Sorry, I'm not clear what you're asking. The above code is an error because the Current property on the array enumerator is not a ref return ("must be an assignable variable" according to the error message, which is not a good one).

Reply:
I assumed that we treat arrays differently and use the indexer (just like the non-ref case)
so we could actually lower that to `for (int i = 0; i < x.Length; i++) { ref var v = ref x[i]; }` so it wouldn't be an error.

Reply:
It can be added later. May need to think about array variance, etc...
It is not a part of this change.

Reply:
could have been a part of this "feature".. honestly that was my first expectation from this. anyways, I think, at least, the error message could be more specific there.